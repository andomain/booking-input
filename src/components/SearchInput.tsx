import React, { FC, ChangeEvent, useState, useEffect } from 'react';
import axios, { AxiosResponse } from 'axios';

import { SearchResponse, API_RESPONSE } from '../types';
import SearchResult from './SearchResult';

interface SearchInputProps {
  id: string;
  name: string;
  label: string;
  placeholder: string;
  onSelect: (data: SearchResponse | null) => void;
}

const API_BASE = 'https://www.rentalcars.com';
const MIN_SEARCH_LENGTH = 2;
const API_RESULT_LIMIT = 6;

export const fetchResults = async (
  searchTerm: string,
  limit: number,
): Promise<AxiosResponse<API_RESPONSE>> =>
  axios.get(
    `${API_BASE}/FTSAutocomplete.do?solrIndex=fts_en&solrRows=${limit}&solrTerm=${searchTerm}`,
  );

const SearchInput: FC<SearchInputProps> = ({
  id,
  name,
  label,
  placeholder,
  onSelect,
}) => {
  const [typedValue, setTypedValue] = useState('');
  const [selectedValue, setSelectedValue] = useState('');
  const [displaySelected, setDisplaySelected] = useState(false);
  const [results, setResults] = useState<null | SearchResponse[]>(null);
  const [error, setError] = useState('');

  useEffect(() => {
    setError('');
    if (typedValue.length >= MIN_SEARCH_LENGTH) {
      getResults(typedValue);
    } else {
      setResults(null);
    }
  }, [typedValue]);

  const getResults = async (
    searchTerm: string,
    limit = API_RESULT_LIMIT,
  ): Promise<void> => {
    const { data, status } = await fetchResults(searchTerm, limit);

    if (status === 200) {
      if (data.results.numFound > 0) {
        setResults(data.results.docs);
      } else {
        setResults([]);
      }
    } else {
      setError('Something went wrong');
    }
  };

  const typeHandler = (e: ChangeEvent<HTMLInputElement>): void => {
    const {
      target: { value },
    } = e;
    setTypedValue(value);
  };

  const focusInput = () => setDisplaySelected(false);

  const selectHandler = (displayName: string, data: SearchResponse): void => {
    setDisplaySelected(true);
    setSelectedValue(displayName);
    onSelect(data);
  };

  const reset = (): void => {
    setDisplaySelected(false);
    setSelectedValue('');
    onSelect(null);
  };

  return (
    <div className="SearchInput">
      <label className="SearchInput__Label" htmlFor={id}>
        {label}
      </label>
      <input
        className="SearchInput__Input"
        id={id}
        name={name}
        onClick={focusInput}
        onChange={typeHandler}
        type="text"
        placeholder={placeholder}
        value={displaySelected ? selectedValue : typedValue}
      />
      {error && <div className="SearchInput__Error">{error}</div>}
      <div className="SearchInput__Results">
        <ul>
          {results !== null &&
            results.length > 0 &&
            results.map(result => (
              <SearchResult
                key={result.placeKey}
                data={result}
                onSelect={selectHandler}
              />
            ))}
          {results !== null && results.length === 0 && (
            <li className="SearchResult SearchResult--none" onClick={reset}>
              No results found
            </li>
          )}
        </ul>
      </div>
    </div>
  );
};

export default SearchInput;
