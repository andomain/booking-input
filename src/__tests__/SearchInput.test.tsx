import React from 'react';
import { HTMLAttributes, mount, ReactWrapper } from 'enzyme';

import SearchInput from '../components/SearchInput';

const TEST = {
  ID: 'test-id',
  NAME: 'test-name',
  LABEL: 'test-label',
  PLACEHOLD: 'test-placehold',
};

describe('<SearchInput />', () => {
  let wrapper: ReactWrapper;
  let input: ReactWrapper<HTMLAttributes>;
  let label: ReactWrapper<HTMLAttributes>;

  const mockSelect = jest.fn();

  beforeEach(() => {
    wrapper = mount(
      <SearchInput
        id={TEST.ID}
        name={TEST.NAME}
        label={TEST.LABEL}
        placeholder={TEST.PLACEHOLD}
        onSelect={mockSelect}
      />,
    );
    input = wrapper.find('input');
    label = wrapper.find('label');
  });

  afterEach(() => jest.clearAllMocks());

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('contains a text <input />', () => {
    expect(input.length).toEqual(1);
    expect(input.props().type).toEqual('text');
  });

  it('should display a label', () => {
    expect(label.text()).toEqual(TEST.LABEL);
  });

  it('should display a placeholder', () => {
    expect(input.props().placeholder).toEqual(TEST.PLACEHOLD);
  });

  describe('Accessibility', () => {
    it('should have an id & corresponding label for screenreaders', () => {
      expect(input.props().id).toEqual(TEST.ID);
      expect(label.props().htmlFor).toEqual(TEST.ID);
    });
  });

  // TODO: Get these tests to pass.
  // Ran out of time
  describe('Search functionality', () => {
    describe('Typing', () => {
      it('should not display any results when one alphanumeric character is typed', async () => {
        const getResults = jest.fn();

        input.simulate('change', {
          preventDefault() {},
          target: { value: 'M' },
        });

        expect(getResults).not.toHaveBeenCalled();
      });

      it('should search/display results when 2+ alphanumeric characters typed', async () => {
        const getResults = jest.fn();

        input.simulate('change', {
          preventDefault() {},
          target: { value: 'Man' },
        });

        expect(getResults).toHaveBeenCalled();
      });
    });

    it('should remove the search results when characters are deleted leaving <2', async () => {
      const getResults = jest.fn();

      input.simulate('change', {
        preventDefault() {},
        target: { value: 'Man' },
      });

      expect(getResults).toHaveBeenCalledTimes(1);
      input.simulate('change', {
        preventDefault() {},
        target: { value: 'M' },
      });
      expect(getResults).toHaveBeenCalledTimes(1);
    });
  });
});
