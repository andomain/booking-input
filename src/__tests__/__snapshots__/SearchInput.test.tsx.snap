// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> should render correctly 1`] = `
<SearchInput
  id="test-id"
  label="test-label"
  name="test-name"
  onSelect={[MockFunction]}
  placeholder="test-placehold"
>
  <div
    className="SearchInput"
  >
    <label
      className="SearchInput__Label"
      htmlFor="test-id"
    >
      test-label
    </label>
    <input
      className="SearchInput__Input"
      id="test-id"
      name="test-name"
      onChange={[Function]}
      onClick={[Function]}
      placeholder="test-placehold"
      type="text"
      value=""
    />
    <div
      className="SearchInput__Results"
    >
      <ul />
    </div>
  </div>
</SearchInput>
`;
